src/goto-programs/abstract-interpretation/gcse.cpp

```diff
@@ -0,0 +1,468 @@
+    break;
+  case DEAD:
+    havoc_symbol(to_code_dead2t(instruction.code).value);
+    break;
+  case RETURN:
+  {
+    const code_return2t &cr = to_code_return2t(instruction.code);
+    make_expression_available(cr.operand);
+    break;
+  }
+  case FUNCTION_CALL:
+  {
+    const code_function_call2t &func =
+      to_code_function_call2t(instruction.code);
+
+    // each operand should be available now (unless someone is doing a sideeffect)
+#if 0
+    // Skip functions for now, the abstract interpreter is not context-aware
+    // so it can't deal with function parameters properly.
+    // Problematic benchmark: https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/raw/main/c/reducercommutativity/sum05-2.yml
+    for(const expr2tc &e : func.operands)
+      make_expression_available(e);
+#endif
+    if(func.ret)
+    {
+      havoc_expr(func.ret, to);
+      make_expression_available(func.ret);
+    }
+
+    break;
+  }
+  default:;
+  }
+}
+
+void cse_domaint::output(std::ostream &out) const
+{
+  if(is_bottom())
+  {
+    out << "BOTTOM\n";
+    return;
+  }
+
+  out << "{";
+  for(const auto &e : available_expressions)
+  {
+    out << e->pretty(0) << "\n";
+  }
+  out << "}";
+}
+
+bool cse_domaint::merge(
+  const cse_domaint &b,
+  goto_programt::const_targett from,
+  goto_programt::const_targett to)
+{
+  /* This analysis is supposed to be used for CFGs.
+   * Since we do not have a CFG GOTO abstract interpreter we
+   * simulate one by just passing through common instructions
+   * and only doing intersections at target destinations */
+
+  if(!(to->is_target() || from->is_function_call()) || is_bottom())
+  {
+    bool changed = available_expressions != b.available_expressions;
+    available_expressions = b.available_expressions;
+    return changed;
+  }
+
+  size_t size_before_intersection = available_expressions.size();
+  for(auto it = available_expressions.begin();
+      it != available_expressions.end();
+      it++)
+  {
+    if(!b.available_expressions.count(*it))
+    {
+      it = available_expressions.erase(it);
+      if(it == available_expressions.end())
+        break;
+    }
+  }
+
+  return size_before_intersection != available_expressions.size();
+}
+
+void cse_domaint::make_expression_available(const expr2tc &E)
+{
+  if(!E)
+    return;
+
+  // Logical expressions can be short-circuited
+  // So, only the first operand will be available
+  if(is_and2t(E) || is_or2t(E))
+  {
+    return;
+  }
+
+  if(is_if2t(E))
+  {
+    return;
+  }
+  // Skip sideeffects
+  if(is_sideeffect2t(E))
+    return;
+
+  // TODO: I hate floats
+  if(is_floatbv_type(E))
+    return;
+
+  // Did we check it already?
+  if(available_expressions.count(E))
+    return;
+
+  // Let's recursively make it available!
+  E->foreach_operand(
+    [this](const expr2tc &e) { make_expression_available(e); });
+
+  // TODO: LHS members should always be recomputed
+  if(is_with2t(E) || is_member2t(E) || is_dereference2t(E))
```

Comment:
I don't understand the comment: What does "recomputed" mean here?

Reply:
Recomputed is my way of saying that something is technically available but we should recompute it anyway. For example:

```
int *arr = &some_arr[0];
arr[0] 42; // *(arr + 0) = 42
```

With we cache `*(arr + 0)` into some variable. The goto convertion will generate an invalid expression:

```
int *arr = &some_arr[0];
int deref = *(arr + 0);
deref = 42;
```

This is correct for RHS though, unless there is an assignment on it.